audition/regrets.py

```diff
@@ -55,13 +57,132 @@ def regrets_for_rule(
+            regret_metric (string) The metric (i.e. precision@) to calculate
+                regrets against
+            regret_parameter (string) The metric parameter (i.e. 100_abs) to
+                calculate regrets against
+
+        Returns: (pandas.DataFrame) A dataframe with columns 'regret',
+            'pct_of_time', and 'selection_rule'
+        """
+        accumulator = list()
+        for selection_rule in bound_selection_rules:
+            regrets = self.regret_calculator.regrets_for_rule(
+                selection_rule,
+                model_group_ids,
+                train_end_times,
+                regret_metric,
+                regret_parameter
+            )
+            plot_min, plot_max = self.plot_bounds(regret_metric, regret_parameter)
+            plot_tick_dist = self.plot_tick_dist(plot_min, plot_max)
+            for regret_threshold in numpy.arange(plot_min, plot_max, plot_tick_dist):
+                pct_of_time = numpy.mean([1 if regret < regret_threshold else 0 for regret in regrets])
+                accumulator.append({
+                    'regret': regret_threshold,
+                    'pct_of_time': pct_of_time,
+                    'selection_rule': selection_rule.descriptive_name,
+                })
+        return pandas.DataFrame.from_records(accumulator)
+
+    def plot_all_selection_rules(
+        self,
+        bound_selection_rules,
+        model_group_ids,
+        train_end_times,
+        regret_metric,
+        regret_parameter
+    ):
+        """Plot the regrets of all given selection rules
+
+        Args:
+            bound_selection_rules (list of .selection_rules.BoundSelectionRule)
+               Selection rule/parameter combinations to plot
+            model_group_ids (list of integers) The model group ids to include
+                in calculating the data
+            train_end_times (list of timestamps) The timestamps to include in
+                calculating the data
+            regret_metric (string) The metric (i.e. precision@) to calculate
+                regrets against
+            regret_parameter (string) The metric parameter (i.e. 100_abs) to
+                calculate regrets against
+        """
+        df_regrets = self.create_plot_dataframe(
+            bound_selection_rules,
+            model_group_ids,
+            train_end_times,
+            regret_metric,
+            regret_parameter
+        )
+        cat_col = 'selection_rule'
+        plt_title = 'Fraction of models X pp worse than best {} {} next time'.format(regret_metric, regret_parameter)
+
+        plot_cats(
+            frame=df_regrets,
+            x_col='regret',
+            y_col='pct_of_time',
+            cat_col=cat_col,
+            grp_col='selection_rule',
+            title=plt_title,
+            x_label='decrease in {} next time from best model'.format(regret_metric),
+            y_label='fraction of models',
+            x_ticks=numpy.arange(0, 1.1, 0.1)
```

Comment:
These ticks assume a 0-1 range, right? Also, the x-label of `decrease...` should probably be something like `distance from best`